### PR TITLE
Add ranges field to PSP group Validate error details

### LINKED ISSUE #60847: PSP: improve error message for invalid supplementalGroups/fsGroup
**Is this a BUG REPORT or FEATURE REQUEST?**:
This is request for improvements.

**What happened**:
When PodSecurityPolicy provider validates `SecurityContext.SupplementalGroups` or `fsGroup` fields of a pod, it shows an error like "999 is not an allowed group" if the group doesn't belong to the range(s).

**What you expected to happen**:
It should include to the error message also the allowed values -- the ranges against what we were validating this group. This will simplify user experience and debugging.

**How to reproduce it (as minimally and precisely as possible)**:
- run kubernetes cluster with PSP support
- create a PSP that allows fsGroup/SupplementalGroups in some ranges
- grant to a user access to this PSP
- try to create a pod that has fsGroup/SupplementalGroups outside of the allowed ranges

**Anything else we need to know?**:
In order to improve the current behavior, we need to modify `Validate()` function here: https://github.com/kubernetes/kubernetes/blob/48a7048d9819ee516b3cfdf5c458b5815b640138/pkg/security/podsecuritypolicy/group/mustrunas.go#L71-L76

Reported-by: @dmage 

@kubernetes/sig-auth-feature-requests 
/kind help

### PR DESCRIPTION
**Pull request by lnsp**
**What this PR does / why we need it**: When the PodSecurityPolicy provider validation of a `fsGroup` fails, it throws a very non-descriptive error that redundantly announces that the group ID is not allowed. Since the group ID is already contained in the error stack, this is not required. This PR fixes it by storing the allowed ranges in the error details and therefore improves the debugging experience.

**Which issue(s) this PR fixes**:
Fixes #60847 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by php-coder — CHANGES_REQUESTED** [review]
Thank you for working on this!

I left a comment about how we can improve this message even more. Also, please, add/update unit tests (see `mustrunas_test.go` file in the same directory).

**Comment by php-coder** [comment]
@lnsp Let me know when you will update this PR and tests, I'll add an ok-to-test label to run all the tests.

**Comment by lnsp** [comment]
@php-coder I've added a field to the test table to verify the error output and extended the existing cases, should I add additional test cases for multiple invalid groups?

File: pkg/security/podsecuritypolicy/group/mustrunas_test.go
Comment (php-coder) [inline pkg/security/podsecuritypolicy/group/mustrunas_test.go]: The code would be simpler if you hold only a single error instead of a list.

File: pkg/security/podsecuritypolicy/group/mustrunas_test.go
Comment (php-coder) [inline pkg/security/podsecuritypolicy/group/mustrunas_test.go]: The message isn't perfect but better than it was. I won't insist on making the ranges more readable, maybe you or other reviewers decide that it's required.

BTW @dmage WDYT do we need to further improve it?

File: pkg/security/podsecuritypolicy/group/mustrunas_test.go
Comment (php-coder) [reply pkg/security/podsecuritypolicy/group/mustrunas_test.go]: P.S. Even not an error but an error message.

**Review by php-coder — CHANGES_REQUESTED** [review]
Some more comments from me. Also, please, squash the commits into single.

**Comment by lnsp** [comment]
@php-coder I changed the tests according to the review, anything missing?

**Comment by php-coder** [comment]
> I changed the tests according to the review, anything missing?

No, it's LGTM now. Thank you!

/lgtm

**Comment by php-coder** [comment]
/assign @pweil- 

PTAL

File: pkg/security/podsecuritypolicy/group/mustrunas.go
Comment (pweil-) [inline pkg/security/podsecuritypolicy/group/mustrunas.go]: @php-coder do we have a bug here?  Should it return early if it finds valid?  The scenario I'm thinking is that if you have multiple ranges defined and it matches on the second range won't it still return an error.  Please add a test for this and check the other ranged strategies.

We also don't seem to be testing that we get the expected number of errors on multiple non-matches.

**Review by pweil- — CHANGES_REQUESTED** [review]

File: pkg/security/podsecuritypolicy/group/mustrunas.go
Comment (php-coder) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: @pweil- Good catch! Look like a bug, indeed. Should it be fixed separately? In this case, it will be easier to backport it if needed and it won't block this PR. WDYT?

Comment (pweil-) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: Yes, it isn't directly related to the initial intent of this PR so I'm ok with an issue and follow up now that it's identified.

File: pkg/security/podsecuritypolicy/group/mustrunas_test.go
Comment (php-coder) [inline pkg/security/podsecuritypolicy/group/mustrunas_test.go]: nit: instead of `errs[0].Error()` you can use `errs[0]` and it will work because of `%v`.

File: pkg/security/podsecuritypolicy/group/mustrunas.go
Comment (php-coder) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: @lnsp Lennart, we need your opinion here before proceed. Please, let us know what option is preferable to you:
1) you will fix the both issues as part of this PR
2) you will fix the both issues but as 2 separate PRs (in this case the 2nd will depend on 1st)
3) you work only on the current PR and I will fix this issue as part on another PR

Comment (lnsp) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: I would also leave it for another PR since it seems to be a bit out of scope.

Comment (php-coder) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: @lnsp Are you interested in working on it? (it should be simple and I could help on all the way ;)

Comment (lnsp) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: Sure, why not!

**Comment by php-coder** [comment]
/lgtm

@pweil- Could you approve it if you agree for having 2 PRs?

**Review by pweil- — APPROVED** [review]
/lgtm

File: pkg/security/podsecuritypolicy/group/mustrunas.go
Comment (php-coder) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: @pweil- @lnsp I tried to reproduce a bug and couldn't. I looked again to the code and now I think that there is no bug. :)

> Should it return early if it finds valid?

This particular loop iterates over pod's supplementalGroups and we need to sure that all user provided groups are fall in in the allowed ranges.

Comment (pweil-) [reply pkg/security/podsecuritypolicy/group/mustrunas.go]: +1, after a second look I agree.  Thanks for following up.

**Pull request merged**
